File: clients/go/events.go
package sidecar

import (
	"context"
	"errors"

	"github.com/ethereum/go-ethereum/common"
	pb "github.com/zama-ai/sdk/clients/go/internal/gen/zama/sdk/v1alpha1"
)

type SDKEventKind int32
type EventOperation int32
type ShieldPath int32
type ApprovalStep int32

func (kind SDKEventKind) String() string   { return pb.SdkEventKind(kind).String() }
func (kind EventOperation) String() string { return pb.EventOperation(kind).String() }
func (path ShieldPath) String() string     { return pb.ShieldPath(path).String() }
func (step ApprovalStep) String() string   { return pb.ApprovalStep(step).String() }

const (
	EventOperationUnspecified            = EventOperation(pb.EventOperation_EVENT_OPERATION_UNSPECIFIED)
	EventOperationGrantPermit            = EventOperation(pb.EventOperation_EVENT_OPERATION_GRANT_PERMIT)
	EventOperationGrantDelegationPermit  = EventOperation(pb.EventOperation_EVENT_OPERATION_GRANT_DELEGATION_PERMIT)
	EventOperationRegisterPermit         = EventOperation(pb.EventOperation_EVENT_OPERATION_REGISTER_PERMIT)
	EventOperationApproveUnderlying      = EventOperation(pb.EventOperation_EVENT_OPERATION_APPROVE_UNDERLYING)
	EventOperationApproveUnderlyingReset = EventOperation(pb.EventOperation_EVENT_OPERATION_APPROVE_UNDERLYING_RESET)
	EventOperationDelegateDecryption     = EventOperation(pb.EventOperation_EVENT_OPERATION_DELEGATE_DECRYPTION)
	EventOperationFinalizeUnwrap         = EventOperation(pb.EventOperation_EVENT_OPERATION_FINALIZE_UNWRAP)
	EventOperationRevokeDelegation       = EventOperation(pb.EventOperation_EVENT_OPERATION_REVOKE_DELEGATION)
	EventOperationSetOperator            = EventOperation(pb.EventOperation_EVENT_OPERATION_SET_OPERATOR)
	EventOperationShieldTransferAndCall  = EventOperation(pb.EventOperation_EVENT_OPERATION_SHIELD_TRANSFER_AND_CALL)
	EventOperationShieldApproveAndWrap   = EventOperation(pb.EventOperation_EVENT_OPERATION_SHIELD_APPROVE_AND_WRAP)
	EventOperationWrap                   = EventOperation(pb.EventOperation_EVENT_OPERATION_WRAP)
	EventOperationTransfer               = EventOperation(pb.EventOperation_EVENT_OPERATION_TRANSFER)
	EventOperationTransferAndCall        = EventOperation(pb.EventOperation_EVENT_OPERATION_TRANSFER_AND_CALL)
	EventOperationTransferFrom           = EventOperation(pb.EventOperation_EVENT_OPERATION_TRANSFER_FROM)
	EventOperationTransferFromAndCall    = EventOperation(pb.EventOperation_EVENT_OPERATION_TRANSFER_FROM_AND_CALL)
	EventOperationUnwrap                 = EventOperation(pb.EventOperation_EVENT_OPERATION_UNWRAP)
	EventOperationUnwrapAll              = EventOperation(pb.EventOperation_EVENT_OPERATION_UNWRAP_ALL)
	ShieldPathUnspecified                = ShieldPath(pb.ShieldPath_SHIELD_PATH_UNSPECIFIED)
	ShieldPathTransferAndCall            = ShieldPath(pb.ShieldPath_SHIELD_PATH_TRANSFER_AND_CALL)
	ShieldPathApproveAndWrap             = ShieldPath(pb.ShieldPath_SHIELD_PATH_APPROVE_AND_WRAP)
	ApprovalStepUnspecified              = ApprovalStep(pb.ApprovalStep_APPROVAL_STEP_UNSPECIFIED)
	ApprovalStepReset                    = ApprovalStep(pb.ApprovalStep_APPROVAL_STEP_RESET)
	ApprovalStepApprove                  = ApprovalStep(pb.ApprovalStep_APPROVAL_STEP_APPROVE)
)

const (
	SDKEventKindUnspecified            = SDKEventKind(pb.SdkEventKind_SDK_EVENT_KIND_UNSPECIFIED)
	SDKEventEncryptStart               = SDKEventKind(pb.SdkEventKind_SDK_EVENT_KIND_ENCRYPT_START)
	SDKEventEncryptEnd                 = SDKEventKind(pb.SdkEventKind_SDK_EVENT_KIND_ENCRYPT_END)
	SDKEventEncryptError               = SDKEventKind(pb.SdkEventKind_SDK_EVENT_KIND_ENCRYPT_ERROR)
	SDKEventDecryptStart               = SDKEventKind(pb.SdkEventKind_SDK_EVENT_KIND_DECRYPT_START)
	SDKEventDecryptEnd                 = SDKEventKind(pb.SdkEventKind_SDK_EVENT_KIND_DECRYPT_END)
	SDKEventDecryptError               = SDKEventKind(pb.SdkEventKind_SDK_EVENT_KIND_DECRYPT_ERROR)
	SDKEventPermitError                = SDKEventKind(pb.SdkEventKind_SDK_EVENT_KIND_PERMIT_ERROR)
	SDKEventTransactionError           = SDKEventKind(pb.SdkEventKind_SDK_EVENT_KIND_TRANSACTION_ERROR)
	SDKEventShieldSubmitted            = SDKEventKind(pb.SdkEventKind_SDK_EVENT_KIND_SHIELD_SUBMITTED)
	SDKEventTransferSubmitted          = SDKEventKind(pb.SdkEventKind_SDK_EVENT_KIND_TRANSFER_SUBMITTED)
	SDKEventTransferFromSubmitted      = SDKEventKind(pb.SdkEventKind_SDK_EVENT_KIND_TRANSFER_FROM_SUBMITTED)
	SDKEventSetOperatorSubmitted       = SDKEventKind(pb.SdkEventKind_SDK_EVENT_KIND_SET_OPERATOR_SUBMITTED)
	SDKEventApproveUnderlyingSubmitted = SDKEventKind(pb.SdkEventKind_SDK_EVENT_KIND_APPROVE_UNDERLYING_SUBMITTED)
	SDKEventWrapSubmitted              = SDKEventKind(pb.SdkEventKind_SDK_EVENT_KIND_WRAP_SUBMITTED)
	SDKEventUnwrapSubmitted            = SDKEventKind(pb.SdkEventKind_SDK_EVENT_KIND_UNWRAP_SUBMITTED)
	SDKEventFinalizeUnwrapSubmitted    = SDKEventKind(pb.SdkEventKind_SDK_EVENT_KIND_FINALIZE_UNWRAP_SUBMITTED)
	SDKEventDelegationSubmitted        = SDKEventKind(pb.SdkEventKind_SDK_EVENT_KIND_DELEGATION_SUBMITTED)
	SDKEventRevokeDelegationSubmitted  = SDKEventKind(pb.SdkEventKind_SDK_EVENT_KIND_REVOKE_DELEGATION_SUBMITTED)
	SDKEventUnshieldPhase1Submitted    = SDKEventKind(pb.SdkEventKind_SDK_EVENT_KIND_UNSHIELD_PHASE1_SUBMITTED)
	SDKEventUnshieldPhase2Started      = SDKEventKind(pb.SdkEventKind_SDK_EVENT_KIND_UNSHIELD_PHASE2_STARTED)
	SDKEventUnshieldPhase2Submitted    = SDKEventKind(pb.SdkEventKind_SDK_EVENT_KIND_UNSHIELD_PHASE2_SUBMITTED)
)

type SDKEvent struct {
	Kind            SDKEventKind
	Timestamp       float64
	TokenAddress    *common.Address
	SDKOperationID  *string
	DurationMS      *float64
	EncryptedValues []common.Hash
	Result          map[common.Hash]ClearValue
	Error           *SDKError
	Operation       *EventOperation
	TxHash          *common.Hash
	ShieldPath      *ShieldPath
	Step            *ApprovalStep
}
type EventCorrelation struct {
	ContextID   string
	OperationID string
	Sequence    uint64
}
type WalletAccountChanged struct{ Previous, Next *WalletAccount }
type ProgressKind int32

const (
	ProgressKindUnspecified   = ProgressKind(pb.ProgressKind_PROGRESS_KIND_UNSPECIFIED)
	ProgressEncryptComplete   = ProgressKind(pb.ProgressKind_PROGRESS_KIND_ENCRYPT_COMPLETE)
	ProgressTransferSubmitted = ProgressKind(pb.ProgressKind_PROGRESS_KIND_TRANSFER_SUBMITTED)
	ProgressApprovalSubmitted = ProgressKind(pb.ProgressKind_PROGRESS_KIND_APPROVAL_SUBMITTED)
	ProgressShieldSubmitted   = ProgressKind(pb.ProgressKind_PROGRESS_KIND_SHIELD_SUBMITTED)
	ProgressWrapSubmitted     = ProgressKind(pb.ProgressKind_PROGRESS_KIND_WRAP_SUBMITTED)
	ProgressUnwrapSubmitted   = ProgressKind(pb.ProgressKind_PROGRESS_KIND_UNWRAP_SUBMITTED)
	ProgressFinalizing        = ProgressKind(pb.ProgressKind_PROGRESS_KIND_FINALIZING)
	ProgressFinalizeSubmitted = ProgressKind(pb.ProgressKind_PROGRESS_KIND_FINALIZE_SUBMITTED)
)

func (kind ProgressKind) String() string { return pb.ProgressKind(kind).String() }

type OperationProgress struct {
	Kind   ProgressKind
	TxHash *common.Hash
}

// EventHandlers process notifications in delivery order.
// Handlers must honor cancellation; notification errors do not fail SDK operations.
type EventHandlers struct {
	OnEvent                func(context.Context, EventCorrelation, SDKEvent) error
	OnWalletAccountChanged func(context.Context, EventCorrelation, WalletAccountChanged) error
	OnProgress             func(context.Context, EventCorrelation, OperationProgress) error
}

func eventAddress(value []byte) (*common.Address, error) {
	if value == nil {
		return nil, nil
	}
	if len(value) != common.AddressLength {
		return nil, errors.New("invalid event address")
	}
	result := common.BytesToAddress(value)
	return &result, nil
}
func eventHash(value []byte) (*common.Hash, error) {
	if value == nil {
		return nil, nil
	}
	if len(value) != common.HashLength {
		return nil, errors.New("invalid event hash")
	}
	result := common.BytesToHash(value)
	return &result, nil
}
func eventAccount(value *pb.WalletAccount) (*WalletAccount, error) {
	if value == nil {
		return nil, nil
	}
	address, err := eventAddress(value.Address)
	if err != nil {
		return nil, err
	}
	if address == nil {
		return nil, errors.New("missing wallet address")
	}
	return &WalletAccount{Address: *address, ChainID: value.ChainId}, nil
}
func sdkEvent(value *pb.SdkEvent) (SDKEvent, error) {
	if value == nil {
		return SDKEvent{}, errors.New("missing SDK event")
	}
	result := SDKEvent{Kind: SDKEventKind(value.Type), Timestamp: value.Timestamp, SDKOperationID: value.SdkOperationId, DurationMS: value.DurationMs}
	if value.Operation != nil {
		operation := EventOperation(*value.Operation)
		result.Operation = &operation
	}
	if value.ShieldPath != nil {
		path := ShieldPath(*value.ShieldPath)
		result.ShieldPath = &path
	}
	if value.Step != nil {
		step := ApprovalStep(*value.Step)
		result.Step = &step
	}
	var err error
	result.TokenAddress, err = eventAddress(value.TokenAddress)
	if err != nil {
		return result, err
	}
	result.TxHash, err = eventHash(value.TxHash)
	if err != nil {
		return result, err
	}
	if value.Error != nil {
		result.Error = sdkError(value.Error)
	}
	for _, bytes := range value.EncryptedValues {
		hash, err := eventHash(bytes)
		if err != nil {
			return result, err
		}
		if hash == nil {
			return result, errors.New("missing encrypted value")
		}
		result.EncryptedValues = append(result.EncryptedValues, *hash)
	}
	result.Result, err = clearEntries(value.Result)
	return result, err
}

File: clients/rust/src/events.rs
use crate::{Address, B256, ClearValues, SdkError, WalletAccount, async_trait, generated};
use anyhow::{Result, ensure};
use std::fmt;

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum EventEnum<E> {
    Known(E),
    Unknown(i32),
}
impl<E: TryFrom<i32>> EventEnum<E> {
    pub fn from_raw(value: i32) -> Self {
        match E::try_from(value) {
            Ok(known) => Self::Known(known),
            Err(_) => Self::Unknown(value),
        }
    }
}
impl<E: fmt::Debug> fmt::Display for EventEnum<E> {
    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Self::Known(value) => write!(formatter, "{value:?}"),
            Self::Unknown(value) => write!(formatter, "Unknown({value})"),
        }
    }
}
pub type EventKind = EventEnum<generated::SdkEventKind>;

/// May contain decrypted plaintext; select metadata explicitly when writing diagnostics.
#[derive(Clone, Debug, PartialEq)]
pub struct SdkEvent {
    pub kind: EventKind,
    pub timestamp: f64,
    pub token_address: Option<Address>,
    /// ID assigned by the SDK; it can differ from the RPC operation ID.
    pub sdk_operation_id: Option<String>,
    pub duration_ms: Option<f64>,
    pub encrypted_values: Vec<B256>,
    pub result: ClearValues,
    pub error: Option<SdkError>,
    pub operation: Option<EventEnum<generated::EventOperation>>,
    pub tx_hash: Option<B256>,
    pub shield_path: Option<EventEnum<generated::ShieldPath>>,
    pub step: Option<EventEnum<generated::ApprovalStep>>,
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct EventContext {
    pub context_id: String,
    /// Absent for SDK activity outside a unary RPC.
    pub operation_id: Option<String>,
    pub sequence: u64,
}
#[derive(Clone, Debug, PartialEq, Eq)]
pub struct OperationProgress {
    pub kind: EventEnum<generated::ProgressKind>,
    pub tx_hash: Option<B256>,
}
#[derive(Clone, Debug, PartialEq)]
pub enum Notification {
    Lifecycle(Box<SdkEvent>),
    WalletAccountChanged {
        previous: Option<WalletAccount>,
        next: Option<WalletAccount>,
    },
    Progress(OperationProgress),
}
/// Notifications run sequentially. Errors are reported to the sidecar; a panic closes the subscription.
#[async_trait]
pub trait EventHandler: Send + Sync {
    async fn on_notification(
        &self,
        context: EventContext,
        notification: Notification,
    ) -> Result<()>;
}

pub(crate) fn address(bytes: &[u8]) -> Result<Address> {
    ensure!(bytes.len() == 20, "invalid event token address");
    Ok(Address::from_slice(bytes))
}
impl TryFrom<generated::SdkEvent> for SdkEvent {
    type Error = anyhow::Error;
    fn try_from(event: generated::SdkEvent) -> Result<Self> {
        Ok(Self {
            kind: EventKind::from_raw(event.r#type),
            timestamp: event.timestamp,
            token_address: event.token_address.as_deref().map(address).transpose()?,
            sdk_operation_id: event.sdk_operation_id,
            duration_ms: event.duration_ms,
            encrypted_values: event
                .encrypted_values
                .iter()
                .map(|v| crate::types::handle(v))
                .collect::<Result<_>>()?,
            result: crate::types::clear_values(event.result)?,
            error: event.error.map(Into::into),
            operation: event.operation.map(EventEnum::from_raw),
            tx_hash: event
                .tx_hash
                .as_deref()
                .map(crate::types::handle)
                .transpose()?,
            shield_path: event.shield_path.map(EventEnum::from_raw),
            step: event.step.map(EventEnum::from_raw),
        })
    }
}
impl TryFrom<generated::OperationProgress> for OperationProgress {
    type Error = anyhow::Error;
    fn try_from(progress: generated::OperationProgress) -> Result<Self> {
        Ok(Self {
            kind: EventEnum::from_raw(progress.kind),
            tx_hash: progress
                .tx_hash
                .as_deref()
                .map(crate::types::handle)
                .transpose()?,
        })
    }
}
